cmd/registry/cmd/compute-references.go

```diff
@@ -0,0 +1,144 @@
+// Copyright 2021 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"log"
+
+	"github.com/apigee/registry/cmd/registry/core"
+	"github.com/apigee/registry/connection"
+	"github.com/apigee/registry/rpc"
+	"github.com/apigee/registry/server/names"
+	openapi_v2 "github.com/googleapis/gnostic/openapiv2"
+	openapi_v3 "github.com/googleapis/gnostic/openapiv3"
+	"github.com/spf13/cobra"
+	"google.golang.org/protobuf/proto"
+)
+
+func init() {
+	computeCmd.AddCommand(computeReferencesCmd)
+}
+
+var computeReferencesCmd = &cobra.Command{
+	Use:   "references",
+	Short: "Compute references of API specs",
+	Args:  cobra.MinimumNArgs(1),
+	Run: func(cmd *cobra.Command, args []string) {
+		ctx := context.TODO()
+		client, err := connection.NewClient(ctx)
+		if err != nil {
+			log.Fatalf("%s", err.Error())
+		}
+		// Initialize task queue.
+		taskQueue := make(chan core.Task, 1024)
+		workerCount := 64
+		for i := 0; i < workerCount; i++ {
+			core.WaitGroup().Add(1)
+			go core.Worker(ctx, taskQueue)
+		}
+		// Generate tasks.
+		name := args[0]
+		if m := names.SpecRegexp().FindStringSubmatch(name); m != nil {
+			// Iterate through a collection of specs and compute references for each
+			err = core.ListSpecs(ctx, client, m, computeFilter, func(spec *rpc.ApiSpec) {
+				taskQueue <- &computeReferencesTask{
+					ctx:      ctx,
+					client:   client,
+					specName: spec.Name,
+				}
+			})
+			if err != nil {
+				log.Fatalf("%s", err.Error())
+			}
+			close(taskQueue)
+			core.WaitGroup().Wait()
```

Comment:
Should the workers be cleaned up when the regex pattern doesn't match?

Reply:
Do you mean should we shut down the goroutines? I haven't done that because these command handlers immediately exit. But I guess we could move the close and wait outside of the test to do that.

Reply:
I was just observing that the task queue and worker goroutines will be created and not closed if the name arg is invalid, but will be closed if the arg is valid. I'm curious why the name arg is relevant at all in whether or not we clean up those resources explicitly.